File: app/src/main/java/ru/melodin/fast/view/BoundedLinearLayout.java
package ru.melodin.fast.view;

import android.content.*;
import android.content.res.*;
import android.util.*;
import android.widget.*;
import ru.melodin.fast.*;

/**
 * Created by Igor on 06.03.16.
 * Thanks by Chase
 * <p/>
 * http://stackoverflow.com/questions/5875877/setting-a-maximum-width-on-a-viewgroup
 */
public class BoundedLinearLayout extends LinearLayout {

    private int mBoundedWidth;
    private int mBoundedHeight;

    public BoundedLinearLayout(Context context) {
        super(context);
        mBoundedWidth = 0;
        mBoundedHeight = 0;
    }

    public BoundedLinearLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.BoundedView);
        mBoundedWidth = a.getDimensionPixelSize(R.styleable.BoundedView_bounded_width, 0);
        mBoundedHeight = a.getDimensionPixelSize(R.styleable.BoundedView_bounded_height, 0);
        a.recycle();
    }

    public void setMaxWidth(int width) {
        if (mBoundedWidth != width) {
            mBoundedWidth = width;
            requestLayout();
        }
    }

    public int getMaxWidth() {
        return mBoundedWidth;
    }

    public void setMaxHeight(int height) {
        if (mBoundedHeight != height) {
            mBoundedHeight = height;
            requestLayout();
            ;
        }
    }

    public int getMaxHeight() {
        return mBoundedHeight;
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        // Adjust width as necessary
        int measuredWidth = MeasureSpec.getSize(widthMeasureSpec);
        if (mBoundedWidth > 0 && mBoundedWidth < measuredWidth) {
            int measureMode = MeasureSpec.getMode(widthMeasureSpec);
            widthMeasureSpec = MeasureSpec.makeMeasureSpec(mBoundedWidth, measureMode);
        }
        // Adjust height as necessary
        int measuredHeight = MeasureSpec.getSize(heightMeasureSpec);
        if (mBoundedHeight > 0 && mBoundedHeight < measuredHeight) {
            int measureMode = MeasureSpec.getMode(heightMeasureSpec);
            heightMeasureSpec = MeasureSpec.makeMeasureSpec(mBoundedHeight, measureMode);
        }
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }
}

File: app/src/main/java/ru/melodin/fast/view/BoundedLinearLayout.kt
package ru.melodin.fast.view

import android.content.Context
import android.util.AttributeSet
import android.widget.LinearLayout
import ru.melodin.fast.R

class BoundedLinearLayout : LinearLayout {

    private var mBoundedWidth: Int = 0
    private var mBoundedHeight: Int = 0

    var maxWidth: Int
        get() = mBoundedWidth
        set(width) {
            if (mBoundedWidth != width) {
                mBoundedWidth = width
                requestLayout()
            }
        }

    var maxHeight: Int
        get() = mBoundedHeight
        set(height) {
            if (mBoundedHeight != height) {
                mBoundedHeight = height
                requestLayout()
            }
        }

    constructor(context: Context) : super(context) {
        mBoundedWidth = 0
        mBoundedHeight = 0
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        val a = context.obtainStyledAttributes(attrs, R.styleable.BoundedLinearLayout)
        mBoundedWidth = a.getDimensionPixelSize(R.styleable.BoundedLinearLayout_bounded_width, 0)
        mBoundedHeight = a.getDimensionPixelSize(R.styleable.BoundedLinearLayout_bounded_height, 0)
        a.recycle()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        var measureSpec = widthMeasureSpec
        var heightMeasureSpec1 = heightMeasureSpec
        val measuredWidth = MeasureSpec.getSize(measureSpec)
        if (mBoundedWidth in 1 until measuredWidth) {
            val measureMode = MeasureSpec.getMode(measureSpec)
            measureSpec = MeasureSpec.makeMeasureSpec(mBoundedWidth, measureMode)
        }

        val measuredHeight = MeasureSpec.getSize(heightMeasureSpec1)
        if (mBoundedHeight in 1 until measuredHeight) {
            val measureMode = MeasureSpec.getMode(heightMeasureSpec1)
            heightMeasureSpec1 = MeasureSpec.makeMeasureSpec(mBoundedHeight, measureMode)
        }
        super.onMeasure(measureSpec, heightMeasureSpec1)
    }
}
